pkg/build/controller/build/build_controller.go

```diff
@@ -388,34 +409,65 @@ func (bc *BuildController) resolveOutputDockerImageReference(build *buildapi.Bui
 			var ok bool
 			streamName, tag, ok = imageapi.SplitImageStreamTag(streamName)
 			if !ok {
-				return "", fmt.Errorf("the referenced image stream tag is invalid: %s", outputTo.Name)
+				return "", build.Spec.Output.PushSecret, fmt.Errorf("the referenced image stream tag is invalid: %s", outputTo.Name)
 			}
 			tag = ":" + tag
 		}
 		stream, err := bc.imageStreamStore.ImageStreams(namespace).Get(streamName)
 		if err != nil {
 			if errors.IsNotFound(err) {
-				return "", fmt.Errorf("the referenced output image stream %s/%s does not exist", namespace, streamName)
+				return "", build.Spec.Output.PushSecret, fmt.Errorf("the referenced output image stream %s/%s does not exist", namespace, streamName)
 			}
-			return "", fmt.Errorf("the referenced output image stream %s/%s could not be found by build %s/%s: %v", namespace, streamName, build.Namespace, build.Name, err)
+			return "", build.Spec.Output.PushSecret, fmt.Errorf("the referenced output image stream %s/%s could not be found by build %s/%s: %v", namespace, streamName, build.Namespace, build.Name, err)
 		}
 		if len(stream.Status.DockerImageRepository) == 0 {
 			e := fmt.Errorf("the image stream %s/%s cannot be used as the output for build %s/%s because the integrated Docker registry is not configured and no external registry was defined", namespace, outputTo.Name, build.Namespace, build.Name)
 			bc.recorder.Eventf(build, kapi.EventTypeWarning, "invalidOutput", "Error starting build: %v", e)
-			return "", e
+			return "", build.Spec.Output.PushSecret, e
 		}
 		ref = fmt.Sprintf("%s%s", stream.Status.DockerImageRepository, tag)
 	}
-	return ref, nil
+
+	pushSecret = build.Spec.Output.PushSecret
+	if pushSecret == nil {
+		serviceAccount := build.Spec.ServiceAccount
+		if len(serviceAccount) == 0 {
+			serviceAccount = bootstrappolicy.BuilderServiceAccountName
+		}
+		sa, err := bc.kubeClient.Core().ServiceAccounts(build.Namespace).Get(serviceAccount, metav1.GetOptions{})
+		if err != nil {
+			return ref, build.Spec.Output.PushSecret, fmt.Errorf("Error getting push/pull secrets for service account %s/%s: %v", build.Namespace, serviceAccount, err)
```

Comment:
This seems oddly asymmetric.  Missing SA fails, but missing secret doesn't?  Given that the secret isn't required (you could have anonymous push, right?), it seems like this ought not fail either.

I'm guessing that you're trying to handle a race between provisioning the SA and starting the build, but you could end up racing against the secrets just as easily.  Do you really just want to stop the return from new project request until the SAs have secrets?